### PR TITLE
Fix | Increase routing attempt to 10 in netcore for LoginNoFailover and adding routing support to LoginWithFailover

### PR DESCRIPTION
**Pull request by JRahnama**
fixes issue #2872

### CODE REVIEW

**Comment by JRahnama** [comment]
It seems like the Azure triggers are not working to validate this PR.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs [hunk 1]
```diff
@@ -2795,7 +2802,7 @@ internal SqlFedAuthToken GetFedAuthToken(SqlFedAuthInfo fedAuthInfo)
             // Username to use in error messages.
             string username = null;
 
-            var authProvider = _sqlAuthenticationProviderManager.GetProvider(ConnectionOptions.Authentication);
+            SqlAuthenticationProvider authProvider = SqlAuthenticationProvider.GetProvider(ConnectionOptions.Authentication);
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:2805]: With this change, you can now remove the local variable _sqlAuthenticationProviderManager (lines 142 and 499). This change in the netcore code was to eliminate loading of the manager if AAD auth isn't even used and this syncs it to netfx. Even though technically unrelated to the PR description, my thought was to try to sync netfx and netcore and bit more with future code merging in mind.

Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:2805]: I noticed that. I also made a few other changes to align .NET Core and .NET Framework more closely. Should I create a separate PR for those changes, or are you okay with including them in this one?

Comment (David-Engel) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:2805]: I'd just include it here.

**Review by edwardneal — COMMENTED** [review]
Looks good! Between this and the other PRs, I think the only remaining behaviour difference between .NET Core and Framework for this class is TransparentNetworkIPResolution, which is pretty minor.

I'm not sure when exactly this behaviour is triggered - I imagine it'll be an Azure SQL database with geo-replication and a readable secondary replica configured, but haven't tested.

There might be some extra handling in OnEnvChange to adjust. On [lines 2081-2084](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs#L2081-L2084) in the .NET Core version, we throw an exception if we receive the ENV_LOGSHIPNODE EnvChange. This exception isn't thrown in the .NET Framework.

Looking more generally: while there's a RoutingTDSServer object in the TDS.Servers project, I can't see any test coverage for routing. I think the three critical scenarios to test would be zero routing (with ApplicationIntent of ReadOnly), one re-route to another TDSServer, and a routing loop (to verify that the routingAttemps comparison takes effect and doesn't hang the driver) I'm not certain about a fourth scenario, where a TDS server tries to route to itself - I've not tested whether a replica's read-only routing URLs can actually be misconfigured like that.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs (hunk 1 above)
Comment (JRahnama) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs:2803]: done

**Review by David-Engel — APPROVED** [review]

**Review by cheenamalhotra — APPROVED** [review]

**Pull request merged**